File: additive_animations/src/main/java/at/wirecube/additiveanimations/additive_animator/animation_set/view/ViewAnimationState.java
package at.wirecube.additiveanimations.additive_animator.animation_set.view;

import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import at.wirecube.additiveanimations.additive_animator.animation_set.AnimationAction;
import at.wirecube.additiveanimations.additive_animator.animation_set.AnimationState;

public class ViewAnimationState extends AnimationState<View> {

    @NonNull
    private final List<AnimationAction.Animation<View>> mAnimations = new ArrayList<>();
    @Nullable
    private final AnimationStartAction<View> mStartAction;
    @Nullable
    private final AnimationEndAction<View> mEndAction;

    public ViewAnimationState(@NonNull List<ViewAnimation> animations) {
        this(animations, null, null);
    }

    public ViewAnimationState(@NonNull ViewAnimation... animations) {
        this(Arrays.asList(animations), null, null);
    }

    public ViewAnimationState(@NonNull List<ViewAnimation> animations, @Nullable AnimationStartAction<View> startAction) {
        this(animations, startAction, null);
    }

    public ViewAnimationState(ViewAnimation animation, @Nullable AnimationStartAction<View> startAction) {
        this(Arrays.asList(animation), startAction, null);
    }

    public ViewAnimationState(@Nullable AnimationStartAction<View> startAction, @NonNull ViewAnimation... animations) {
        this(Arrays.asList(animations), startAction, null);
    }

    public ViewAnimationState(@NonNull ViewAnimation animation, @Nullable AnimationEndAction<View> endAction) {
        this(Arrays.asList(animation), null, endAction);
    }

    public ViewAnimationState(@NonNull List<ViewAnimation> animations, @Nullable AnimationEndAction<View> endAction) {
        this(animations, null, endAction);
    }

    public ViewAnimationState(@Nullable AnimationEndAction<View> endAction, @NonNull ViewAnimation... animations) {
        this(Arrays.asList(animations), null, endAction);
    }

    public ViewAnimationState(@NonNull ViewAnimation animation, @Nullable AnimationStartAction<View> startAction, @Nullable AnimationEndAction<View> endAction) {
        this(Arrays.asList(animation), startAction, endAction);
    }

    public ViewAnimationState(@Nullable AnimationStartAction<View> startAction, @Nullable AnimationEndAction<View> endAction, @NonNull ViewAnimation... animations) {
        this(Arrays.asList(animations), startAction, endAction);
    }

    public ViewAnimationState(@NonNull List<ViewAnimation> animations, @Nullable AnimationStartAction<View> startAction, @Nullable AnimationEndAction<View> endAction) {
        mAnimations.addAll(animations);
        mStartAction = startAction;
        mEndAction = endAction;
    }

    @Override
    public List<Animation<View>> getAnimations() {
        return mAnimations;
    }

    @Override
    public AnimationStartAction<View> getAnimationStartAction() {
        return mStartAction;
    }

    @Override
    public AnimationEndAction<View> getAnimationEndAction() {
        return mEndAction;
    }
}

File: additive_animations/src/main/java/at/wirecube/additiveanimations/additive_animator/animation_set/view/ViewAnimationState.kt
package at.wirecube.additiveanimations.additive_animator.animation_set.view

import android.view.View
import at.wirecube.additiveanimations.additive_animator.animation_set.AnimationAction
import at.wirecube.additiveanimations.additive_animator.animation_set.AnimationState

class ViewAnimationState : AnimationState<View> {

    private val mAnimations: MutableList<AnimationAction.Animation<View>> = mutableListOf()
    private val mStartAction: AnimationStartAction<View>?
    private val mEndAction: AnimationEndAction<View>?

    constructor(animations: List<ViewAnimation>) : this(animations, null, null)

    constructor(vararg animations: ViewAnimation) : this(animations.toList(), null, null)

    constructor(animations: List<ViewAnimation>, startAction: AnimationStartAction<View>?) : this(animations, startAction, null)

    constructor(animation: ViewAnimation, startAction: AnimationStartAction<View>?) : this(listOf(animation), startAction, null)

    constructor(startAction: AnimationStartAction<View>?, vararg animations: ViewAnimation) : this(animations.toList(), startAction, null)

    constructor(animation: ViewAnimation, endAction: AnimationEndAction<View>?) : this(listOf(animation), null, endAction)

    constructor(animations: List<ViewAnimation>, endAction: AnimationEndAction<View>?) : this(animations, null, endAction)

    constructor(endAction: AnimationEndAction<View>?, vararg animations: ViewAnimation) : this(animations.toList(), null, endAction)

    constructor(animation: ViewAnimation, startAction: AnimationStartAction<View>?, endAction: AnimationEndAction<View>?) : this(listOf(animation), startAction, endAction)

    constructor(startAction: AnimationStartAction<View>?, endAction: AnimationEndAction<View>?, vararg animations: ViewAnimation) : this(animations.toList(), startAction, endAction)

    constructor(
        animations: List<ViewAnimation>,
        startAction: AnimationStartAction<View>?,
        endAction: AnimationEndAction<View>?
    ) {
        mAnimations.addAll(animations)
        mStartAction = startAction
        mEndAction = endAction
    }

    override fun getAnimations(): List<AnimationAction.Animation<View>> = mAnimations

    override fun getAnimationStartAction(): AnimationStartAction<View>? = mStartAction

    override fun getAnimationEndAction(): AnimationEndAction<View>? = mEndAction
}

